python/vast/vast.py

```diff
@@ -0,0 +1,164 @@
+        await proc.communicate()
+        logger.debug(proc.stderr.decode())
+        # TODO: Start a matcher
+        # cmd = CLI(plugins="matcher").matcher().start(match_types="[:string]").test()
+        return proc
+
+    async def republish(self, type: str):
+        """Subscribes to an event type on ingest path and publishes the events
+        directly to the fabric."""
+        callback = lambda line: self.fabric.publish(type, line)
+        await self._live_query(f'#type == "{type}"', callback)
+
+    # Translates an STIX Indicator into a VAST query and publishes the results
+    # as STIX Bundle.
+    async def _on_stix_indicator(self, message):
+        logger.debug(message)
+        indicator = stix2.parse(message)
+        if type(indicator) != stix2.Indicator:
+            logger.warn(f"expected indicator, got {type(indicator)}")
+            return
+        if indicator.pattern_type == "vast":
+            logger.debug(f"got VAST expression {indicator.pattern}")
+        elif indicator.pattern_type == "sigma":
+            logger.debug(f"got Sigma rule {indicator.pattern}")
+        else:
+            logger.warn(f"got unsupported pattern type {indicator.pattern_type}")
+            return
+        results = await self._query(indicator.pattern)
+        if len(results) == 0:
+            logger.info(f"got no results for query: {indicator.pattern}")
+            return
+        bundle = self.stix_bridge.make_sighting(indicator, results)
+        logger.warning(bundle.serialize(pretty=True))
+        await self.fabric.publish("stix.bundle", bundle)
+
+    async def _live_query(self, expression: str, callback):
+        proc = await CLI().export(continuous=True).json(expression).exec()
+        while True:
+            if proc.stdout.at_eof():
+                break
+            line = await proc.stdout.readline()
+            await callback(line)
+        await proc.communicate()
+        if proc.returncode != 0:
+            logger.warning(f"vast exited with non-zero code: {proc.returncode}")
+
+    async def _query(self, expression: str, limit: int = 100):
+        """Executes a VAST and receives results as Arrow Tables."""
+        proc = await CLI().export(max_events=limit).arrow(expression).exec()
+        stdout, stderr = await proc.communicate()
+        logger.debug(stderr.decode())
+        istream = pa.input_stream(io.BytesIO(stdout))
+        num_batches = 0
+        num_rows = 0
+        batches = defaultdict(list)
+        try:
+            while True:
+                reader = pa.ipc.RecordBatchStreamReader(istream)
+                try:
+                    while True:
+                        batch = reader.read_next_batch()
+                        name = vast.utils.arrow.name(batch.schema)
+                        logger.debug(f"got batch of {name}")
+                        num_batches += 1
+                        num_rows += batch.num_rows
+                        batches[name].append(batch)
```

Comment:
This is broken if the you have batches with the same schema name and a different physical layout. I recommend switching to another structure for `batches`. It should still be tables, but ideally a list of them rather than a dict.

Reply:
I fixed this, but it required a workaround of going through the string representation of the schema:

```python
# TODO: this might be slow, but schemas are not
# hashable. The string representation is the next best
# thing to determine uniqueness.
batches[batch.schema.to_string()].append(batch)
```